Better errors with bad/missing identifiers in MBEs

EliseZeroTwo opened the pull request:
This PR implements better error handling around identifiers in MBEs.

It is related to #118295 which is already closed, but it fixes up issues left present by the solution for that issue.

Improved errors:

<details>
<summary>Parenthesised Ident</summary>
<br>

Previously with the code:

```rs
macro_rules! (meow) {
    () => {}
}
```

The compiler would output the following, incorrect error:

```
error: macros that expand to items must be delimited with braces or followed by a semicolon
 --> src/lib.rs:1:14
  |
1 | macro_rules! (meow) {
  |              ^^^^^^
  |
help: change the delimiters to curly braces
  |
1 | macro_rules! {meow} {
  |              ~    ~
help: add a semicolon
  |
1 | macro_rules! (meow); {
  |                    +

error: expected item, found `{`
 --> src/lib.rs:1:21
  |
1 | macro_rules! (meow) {
  |                     ^ expected item
  |
  = note: for a full list of items that can appear in modules, see <https://doc.rust-lang.org/reference/items.html>
```

Now it outputs:

```
error: expected identifier, found `(`
 --> src/lib.rs:1:14
  |
1 | macro_rules! (meow) {
  |              ^ expected identifier
  |
note: try removing the parenthesis around the name for this `macro_rules!`
 --> src/lib.rs:1:15
  |
1 | macro_rules! (meow) {
  |               ^^^^
```

</details>

<details>
<summary>Missing Ident</summary>
<br>
Previously with the code:

```rs
macro_rules! {}
```

The compiler would output the following, incorrect error:

```
error: cannot find macro `macro_rules` in this scope
 --> src/lib.rs:1:1
  |
1 | macro_rules! {}
  | ^^^^^^^^^^^
```

Now it outputs:

```
error: expected identifier, found `{`
 --> src/lib.rs:1:14
  |
1 | macro_rules! {}
  |              ^ expected identifier
  |
  = note: maybe you have forgotten to define a name for this `macro_rules!`
```
</details>

This will interfere with the `stderr` of the test case I introduced in #118928 that has not yet been merged, due to that test case requiring a (templated) parenthesised ident to be present.

It includes regression tests.

🖤

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7210686835/job/19644388165) [(plain)](https://github.com/rust-lang/rust/commit/aeb58d50b49c9f8c974f2a803b2b128e1296e734/checks/19644388165/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=EliseZeroTwo
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_13195137-1796-4faf-9fea-0ff0f2c8378f
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=EliseZeroTwo/fix-macrorules-ident-errors
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_13195137-1796-4faf-9fea-0ff0f2c8378f
GITHUB_REF=refs/pull/118939/merge
GITHUB_REF_NAME=118939/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=eb38f5a15d4f59d2074f968881f78bac3ede4029
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_13195137-1796-4faf-9fea-0ff0f2c8378f
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_13195137-1796-4faf-9fea-0ff0f2c8378f
GITHUB_TRIGGERING_ACTOR=EliseZeroTwo
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118939/merge
GITHUB_WORKFLOW_SHA=eb38f5a15d4f59d2074f968881f78bac3ede4029
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Thu Dec 14 15:05:03 UTC 2023
  network time: Thu, 14 Dec 2023 15:05:03 GMT
  network time: Thu, 14 Dec 2023 15:05:03 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-missing-tools', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: rust.codegen-units-std := 1
---
   Compiling lock_api v0.4.10
   Compiling generic-array v0.14.7
   Compiling parking_lot_core v0.8.6
   Compiling instant v0.1.12
error: expected `!` after `macro_rules`
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro-hack-0.5.20+deprecated/src/lib.rs:346:9
    |
346 |         macro_rules.extend(quote! {
    |
help: add a `!`
    |
    |
346 |         macro_rules!.extend(quote! {

error: expected identifier, found `.`
   --> /cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro-hack-0.5.20+deprecated/src/lib.rs:346:20
    |
    |
346 |         macro_rules.extend(quote! {
    |
    |
    = note: maybe you have forgotten to define a name for this `macro_rules!`
   Compiling thin-vec v0.2.12
   Compiling bitflags v1.3.2
   Compiling arrayvec v0.7.4
   Compiling fastrand v2.0.0
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7211132359/job/19645840840) [(plain)](https://github.com/rust-lang/rust/commit/a51902ad96a618384c94a076f4fb653a15909d41/checks/19645840840/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=EliseZeroTwo
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_36a52330-a2f4-47a6-b08f-8213f2bd1916
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=EliseZeroTwo/fix-macrorules-ident-errors
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_36a52330-a2f4-47a6-b08f-8213f2bd1916
GITHUB_REF=refs/pull/118939/merge
GITHUB_REF_NAME=118939/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=1cecd6cb73620a06dc94ac77a4609658e67e9091
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_36a52330-a2f4-47a6-b08f-8213f2bd1916
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_36a52330-a2f4-47a6-b08f-8213f2bd1916
GITHUB_TRIGGERING_ACTOR=EliseZeroTwo
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118939/merge
GITHUB_WORKFLOW_SHA=1cecd6cb73620a06dc94ac77a4609658e67e9091
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Thu Dec 14 15:39:07 UTC 2023
  network time: Thu, 14 Dec 2023 15:39:07 GMT
  network time: Thu, 14 Dec 2023 15:39:07 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-missing-tools', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: rust.codegen-units-std := 1
---

1 error: expected identifier, found `(`
2   --> $DIR/mbe-parenthesis-ident-error.rs:3:14
3    |
- LL | macro_rules! (meepmeep) { 
+ LL | macro_rules! (meepmeep) {
6    |
7 note: try removing the parenthesis around the name for this `macro_rules!`

8   --> $DIR/mbe-parenthesis-ident-error.rs:3:15
8   --> $DIR/mbe-parenthesis-ident-error.rs:3:15
9    |
- LL | macro_rules! (meepmeep) { 
+ LL | macro_rules! (meepmeep) {
12 
13 error: aborting due to 1 previous error



The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/mbe-parenthesis-ident-error/mbe-parenthesis-ident-error.stderr
To only update this specific test, also pass `--test-args macros/mbe-parenthesis-ident-error.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/mbe-parenthesis-ident-error.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/mbe-parenthesis-ident-error" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/mbe-parenthesis-ident-error/auxiliary"
--- stderr -------------------------------
error: expected identifier, found `(`
##[error]  --> /checkout/tests/ui/macros/mbe-parenthesis-ident-error.rs:3:14
   |
   |
LL | macro_rules! (meepmeep) {
Build completed unsuccessfully in 0:13:23
   |
note: try removing the parenthesis around the name for this `macro_rules!`
  --> /checkout/tests/ui/macros/mbe-parenthesis-ident-error.rs:3:15
  --> /checkout/tests/ui/macros/mbe-parenthesis-ident-error.rs:3:15
   |
LL | macro_rules! (meepmeep) {

error: aborting due to 1 previous error
------------------------------------------

---
--- stderr -------------------------------
error: expected identifier, found `{`
##[error]  --> /checkout/tests/ui/macros/user-defined-macro-rules.rs:5:14
   |
LL | macro_rules! {} // OK, calls the macro defined above
   |
   |
   = note: maybe you have forgotten to define a name for this `macro_rules!`
error: aborting due to 1 previous error
------------------------------------------


```

</details>

EliseZeroTwo commented:
In #62258 `macro_rules` was reallowed as an identifier for MBE definitions. This removes the possibility of proper error checking of MBEs at the parser level. Given that in the merge request @eddyb stated it could be potentially reverted in the future, I was wondering what the next step with this should be. As per the specification it appears to me as if you shouldn't be allowed to define MBEs with the `macro_rules` identifier.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7213128413/job/19652295125) [(plain)](https://github.com/rust-lang/rust/commit/953a1efd4af134a1ef79cf9fb092ccd58133066e/checks/19652295125/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=EliseZeroTwo
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_31d40c0b-b729-4952-855b-c888ba613311
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=EliseZeroTwo/fix-macrorules-ident-errors
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_31d40c0b-b729-4952-855b-c888ba613311
GITHUB_REF=refs/pull/118939/merge
GITHUB_REF_NAME=118939/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=03d6fce0d393c60fea77a195871a2396ce17fa55
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_31d40c0b-b729-4952-855b-c888ba613311
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_31d40c0b-b729-4952-855b-c888ba613311
GITHUB_TRIGGERING_ACTOR=EliseZeroTwo
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118939/merge
GITHUB_WORKFLOW_SHA=03d6fce0d393c60fea77a195871a2396ce17fa55
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Thu Dec 14 18:38:55 UTC 2023
  network time: Thu, 14 Dec 2023 18:38:55 GMT
  network time: Thu, 14 Dec 2023 18:38:55 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-missing-tools', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: rust.codegen-units-std := 1
---
--- stderr -------------------------------
error: expected identifier, found `{`
##[error]  --> /checkout/tests/ui/macros/user-defined-macro-rules.rs:5:14
   |
LL | macro_rules! {} // OK, calls the macro defined above
   |
   |
   = note: maybe you have forgotten to define a name for this `macro_rules!`
error: aborting due to 1 previous error
------------------------------------------


```

</details>

TaKO8Ki commented:
@EliseZeroTwo A test `tests/ui/macros/user-defined-macro-rules.rs` has failed on CI. Could you fix it?

TaKO8Ki commented:
@rustbot author

EliseZeroTwo commented:
@TaKO8Ki I have re-implemented the special casing of `macro_rules` not being allowed as an identifier for MBE definitions, and fixed the previously failing test that needed to be adjusted as a result of it (the failing test was also added in #62258 due to the change in it). All tests pass now.

TaKO8Ki commented on compiler/rustc_parse/src/parser/item.rs:
```diff
@@ -2070,7 +2073,39 @@ impl<'a> Parser<'a> {
+                    ) => {
+                        err.span_note(
+                            span,
+                            "try removing the parenthesis around the name for this `macro_rules!`",
```
I think it's better to suggest removing the parenthesis with `span_suggestion` or `span_suggestion_verbose`.

EliseZeroTwo replied:
Thank you, that does seem a lot better, I moved the span to include the parentheses now too.

TaKO8Ki commented on compiler/rustc_parse/src/parser/item.rs:
```diff
@@ -2070,7 +2073,39 @@ impl<'a> Parser<'a> {
+                        );
+                    }
+                    _ => {
+                        err.note(fluent::parse_maybe_missing_macro_rules_name);
```
I understand that you've taken over the previous implementation of `MaybeMissingMacroRulesName`, but using `help` here might be a better option.

```suggestion
                        err.help(fluent::parse_maybe_missing_macro_rules_name);
```

EliseZeroTwo committed:
fix: better error cases with bad/missing identifiers in MBEs

EliseZeroTwo commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7276935540/job/19827832618) [(plain)](https://github.com/rust-lang/rust/commit/661b30a6425249cc1619c3ab2c77636d913e4f19/checks/19827832618/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=EliseZeroTwo
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_2d276083-259d-4c29-9f30-48bb892048ef
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=EliseZeroTwo/fix-macrorules-ident-errors
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_2d276083-259d-4c29-9f30-48bb892048ef
GITHUB_REF=refs/pull/118939/merge
GITHUB_REF_NAME=118939/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=1a2bd6ee60e2e03fe088ec1ad291e952879d12e6
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_2d276083-259d-4c29-9f30-48bb892048ef
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_2d276083-259d-4c29-9f30-48bb892048ef
GITHUB_TRIGGERING_ACTOR=EliseZeroTwo
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118939/merge
GITHUB_WORKFLOW_SHA=1a2bd6ee60e2e03fe088ec1ad291e952879d12e6
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Dec 20 14:46:15 UTC 2023
  network time: Wed, 20 Dec 2023 14:46:15 GMT
  network time: Wed, 20 Dec 2023 14:46:15 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-missing-tools', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: rust.codegen-units-std := 1
---

---- [ui] tests/ui/macros/issue-118786.rs stdout ----
diff of stderr:

- error: macros that expand to items must be delimited with braces or followed by a semicolon
+ error: expected identifier, found `(`
3    |
3    |
4 LL |         macro_rules! $macro_name {
-    |                      ^^^^^^^^^^^
+    |                      ^^^^^^^^^^^ expected identifier
6    |
- help: change the delimiters to curly braces
- help: change the delimiters to curly braces
+ help: try removing the parenthesis around the name for this `macro_rules!`
8    |
- LL |         macro_rules! {} {
-    |                      ~          +
- help: add a semicolon
-    |
- LL |         macro_rules! $macro_name; {
-    |                                 +
+ LL |         macro_rules! meow {
15 
15 
- error: macro expansion ignores token `{` and any following
-    |
-    |
- LL |         macro_rules! $macro_name {
- ...
- ...
- LL | make_macro!((meow));
-    | ------------------- caused by the macro expansion here
-    |
-    = note: the usage of `make_macro!` is likely invalid in item context
- error: cannot find macro `macro_rules` in this scope
-   --> $DIR/issue-118786.rs:7:9
-    |
-    |
- LL |         macro_rules! $macro_name {
- ...
- ...
- LL | make_macro!((meow));
-    |
-    |
- note: maybe you have forgotten to define a name for this `macro_rules!`
-    |
-    |
- LL |         macro_rules! $macro_name {
- ...
- ...
- LL | make_macro!((meow));
-    = note: this error originates in the macro `make_macro` (in Nightly builds, run with -Z macro-backtrace for more info)
- 
- error: aborting due to 3 previous errors
+ error: aborting due to 1 previous error
---
To only update this specific test, also pass `--test-args macros/issue-118786.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/macros/issue-118786.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/issue-118786" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/macros/issue-118786/auxiliary" "--crate-type" "lib" "-O" "-C" "debug-assertions=yes"
--- stderr -------------------------------
error: expected identifier, found `(`
##[error]  --> /checkout/tests/ui/macros/issue-118786.rs:7:22
   |
   |
LL |         macro_rules! $macro_name {
   |                      ^^^^^^^^^^^ expected identifier
   |
help: try removing the parenthesis around the name for this `macro_rules!`
   |
   |
LL |         macro_rules! meow {

error: aborting due to 1 previous error
------------------------------------------



---- [ui] tests/ui/macros/user-defined-macro-rules.rs stdout ----
diff of stderr:

10 LL | macro_rules! {}
11    |              ^ expected identifier
12    |
-    = note: maybe you have forgotten to define a name for this `macro_rules!`
+    = help: maybe you have forgotten to define a name for this `macro_rules!`
15 error: aborting due to 2 previous errors
16 


---
--- stderr -------------------------------
error: user-defined macros may not be named `macro_rules`
##[error]  --> /checkout/tests/ui/macros/user-defined-macro-rules.rs:3:14
   |
LL | macro_rules! macro_rules { () => {} } //~ ERROR: user-defined macros may not be named `macro_rules`

error: expected identifier, found `{`
##[error]  --> /checkout/tests/ui/macros/user-defined-macro-rules.rs:5:14
   |
   |
LL | macro_rules! {} //~ ERROR: expected identifier, found `{`
   |
   |
   = help: maybe you have forgotten to define a name for this `macro_rules!`
error: aborting due to 2 previous errors
------------------------------------------


```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7277883376/job/19830882710) [(plain)](https://github.com/rust-lang/rust/commit/db53e3fce8095dd1b6555b64a24850740f277679/checks/19830882710/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=EliseZeroTwo
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_633851e6-cba0-4b28-bd2e-e63a51f6af95
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=EliseZeroTwo/fix-macrorules-ident-errors
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_633851e6-cba0-4b28-bd2e-e63a51f6af95
GITHUB_REF=refs/pull/118939/merge
GITHUB_REF_NAME=118939/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=e31b303d3f7e7879688d9054e9df117b3c8efcb8
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_633851e6-cba0-4b28-bd2e-e63a51f6af95
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_633851e6-cba0-4b28-bd2e-e63a51f6af95
GITHUB_TRIGGERING_ACTOR=EliseZeroTwo
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118939/merge
GITHUB_WORKFLOW_SHA=e31b303d3f7e7879688d9054e9df117b3c8efcb8
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Dec 20 16:07:02 UTC 2023
  network time: Wed, 20 Dec 2023 16:07:02 GMT
  network time: Wed, 20 Dec 2023 16:07:02 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-missing-tools', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: rust.codegen-units-std := 1
---

10 LL | macro_rules! {}
11    |              ^ expected identifier
12    |
-    = note: maybe you have forgotten to define a name for this `macro_rules!`
+    = help: maybe you have forgotten to define a name for this `macro_rules!`
15 error: aborting due to 2 previous errors
16 


---
--- stderr -------------------------------
error: user-defined macros may not be named `macro_rules`
##[error]  --> /checkout/tests/ui/macros/user-defined-macro-rules.rs:3:14
   |
LL | macro_rules! macro_rules { () => {} } //~ ERROR: user-defined macros may not be named `macro_rules`

error: expected identifier, found `{`
##[error]  --> /checkout/tests/ui/macros/user-defined-macro-rules.rs:5:14
Build completed unsuccessfully in 0:13:32
Build completed unsuccessfully in 0:13:32
   |
LL | macro_rules! {} //~ ERROR: expected identifier, found `{`
   |
   |
   = help: maybe you have forgotten to define a name for this `macro_rules!`
error: aborting due to 2 previous errors
------------------------------------------


```

</details>

EliseZeroTwo commented:
@rustbot author

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7280017058/job/19837541714) [(plain)](https://github.com/rust-lang/rust/commit/4c4b15529a958277011fa6e7f8ede32c498df081/checks/19837541714/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=EliseZeroTwo
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_a2812909-2e2b-495f-9c7b-66de4a35529e
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=EliseZeroTwo/fix-macrorules-ident-errors
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_a2812909-2e2b-495f-9c7b-66de4a35529e
GITHUB_REF=refs/pull/118939/merge
GITHUB_REF_NAME=118939/merge
GITHUB_REF_PROTECTED=false
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=67e7a5f575a39f9084b04efd9d1b4cb62568d2d0
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_a2812909-2e2b-495f-9c7b-66de4a35529e
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_a2812909-2e2b-495f-9c7b-66de4a35529e
GITHUB_TRIGGERING_ACTOR=EliseZeroTwo
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118939/merge
GITHUB_WORKFLOW_SHA=67e7a5f575a39f9084b04efd9d1b4cb62568d2d0
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Dec 20 19:40:07 UTC 2023
  network time: Wed, 20 Dec 2023 19:40:07 GMT
  network time: Wed, 20 Dec 2023 19:40:07 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-missing-tools', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: rust.codegen-units-std := 1
---

1 error: user-defined macros may not be named `macro_rules`
2   --> $DIR/user-defined-macro-rules.rs:3:14
3    |
- LL | macro_rules! macro_rules { () => {} } 
+ LL | macro_rules! macro_rules { () => {} }
6 
7 error: expected identifier, found `{`


---
error: user-defined macros may not be named `macro_rules`
##[error]  --> /checkout/tests/ui/macros/user-defined-macro-rules.rs:3:14
   |
Build completed unsuccessfully in 0:13:17
LL | macro_rules! macro_rules { () => {} }

error: expected identifier, found `{`
##[error]  --> /checkout/tests/ui/macros/user-defined-macro-rules.rs:6:14
   |
   |
LL | macro_rules! {}
   |              ^ expected identifier
   |
   = help: maybe you have forgotten to define a name for this `macro_rules!`
error: aborting due to 2 previous errors
------------------------------------------


```

</details>

EliseZeroTwo commented:
@rustbot review

cjgillot commented on compiler/rustc_parse/src/parser/item.rs:
```diff
@@ -2066,7 +2069,42 @@ impl<'a> Parser<'a> {
+        let ident = match self.parse_ident() {
+            Ok(ident) => ident,
+            Err(mut err) => {
+                match (
+                    &self.token.kind,
+                    self.look_ahead(1, |token| token.ident()),
+                    self.look_ahead(2, |token| {
+                        (token.kind == TokenKind::CloseDelim(Delimiter::Parenthesis))
+                            .then_some(token.span)
+                    }),
+                ) {
+                    (
+                        TokenKind::OpenDelim(Delimiter::Parenthesis),
+                        Some((Ident { name, .. }, _)),
+                        Some(closing_span),
+                    ) => {
```
Could you prefer a let chain here?

EliseZeroTwo replied:
Sure, I changed it to a let chain, it seems quite a bit cleaner. Thank you for the suggestion!

cjgillot reviewed (commented):
r=me with one nit

petrochenkov commented:
This is a breaking language change, if I understand correctly.

EliseZeroTwo commented:
> This is a breaking language change, if I understand correctly.

As documentation seems to indicate that this is how it is meant to be, I did not think it was. I took a look before submitting this and also could not find published software using this, but maybe testing the change against crates.io would be a good idea anyways?

It does undo the change you implemented in #62258 which reallows `macro_rules` as an identifier for macros.

petrochenkov commented:
>this is how it is meant to be

Well, not by me, at least.
`macro_rules` is reserved as minimally as possible, because the eventual goal was always retiring it in favor of `macro` - it isn't strictly necessary to reserve `macro_rules` in this context, so it's not reserved (that's similar to other weak keywords like `union`).

In any case, such changes should go through the language team.
(And I'm personally against it going, because the only benefit here is a minor diagnostic improvement.)

bors commented:
:umbrella: The latest upstream changes (presumably #119146) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

apiraino commented:
I'll try to push this conversation forward by nominating for T-lang discussion, keeping in mind some [raised concerns](https://github.com/rust-lang/rust/pull/118939#issuecomment-1868605386).

@rustbot I-lang-nominated
